notebooks/linear_regression.livemd

```diff
@@ -0,0 +1,610 @@
+
+```elixir
+{corr_size, _} = Nx.shape(correlation)
+correlation_list = Nx.to_flat_list(correlation)
+
+names = [
+  "Bedrooms per rooms",
+  "Households",
+  "Housing median age",
+  "Latitude",
+  "Longitude",
+  "Median income",
+  "Population",
+  "Population per family",
+  "Rooms per family",
+  "Total bedrooms",
+  "Total rooms",
+  "Ocean proximity",
+  "Median house value"
+]
+
+corr_to_plot =
+  Explorer.DataFrame.new(
+    x: List.flatten(List.duplicate(names, corr_size)),
+    y: List.flatten(for name <- names, do: List.duplicate(name, corr_size)),
+    corr_val: correlation_list
+  )
+
+Vl.new(
+  title: [
+    text: "Correlation Matrix for California Housing",
+    offset: 20
+  ],
+  width: 630,
+  height: 630
+)
+|> Vl.data_from_values(corr_to_plot)
+|> Vl.mark(:rect)
+|> Vl.encode_field(:x, "x", type: :nominal, title: "")
+|> Vl.encode_field(:y, "y", type: :nominal, title: "")
+|> Vl.encode_field(:color, "corr_val", type: :quantitative, scale: [scheme: :viridis])
+|> Vl.config(axis: [grid: true, tickband: :extent])
+```
+
+We can spot that _median_house_value_ is strongly correlated with _median_income_. It's pretty straightforward, the more money you have, the more expensive house you can buy. Non-obvious is a negative correlation with _bedrooms_per_rooms_. But it also can be explained. Bedrooms are the most crucial rooms in the house. Firstly, you need to guarantee that you have a house with enough bedrooms. If this condition is satisfied, then you can focus on "additional rooms" like a chill room, cabinets and so on. So if you buy a house with more additional rooms, then you decrease the ratio.
+
+<!-- livebook:{"break_markdown":true} -->
+
+Now we are ready to train a model for the _median_house_value_ prediction. We will use linear regression. In the first step, we create the model by calling the `fit` function.
+
+```elixir
+model = Scholar.Linear.LinearRegression.fit(x_train, y_train)
+```
+
+Now we can predict the values for the test set and measure the error of our prediction. We will calculate root mean square error (RMSE) and mean absolute error (MAE).
+
+```elixir
+defmodule RMSE do
+  import Nx.Defn
+
+  defn root_mean_square_error(y_true, y_pred) do
+    diff = y_true - y_pred
+    (diff * diff) |> Nx.mean() |> Nx.sqrt()
+  end
+end
+```
+
+```elixir
+predictions = Scholar.Linear.LinearRegression.predict(model, x_test)
+rmse = RMSE.root_mean_square_error(y_test, predictions)
+mae = Scholar.Metrics.mean_absolute_error(y_test, predictions)
+{rmse, mae}
+```
+
+Ok, but is it a good or poor estimation? Huh, check the mean value of the target and then compare it to the value of errors.
+
+```elixir
+Nx.mean(y)
+```
+
+For such a simple model as linear regression, it seems to be a pretty good result. But there is space to improve this result. You can, for example, add some additional features like the logarithm of the population. In the future, you will be able to try more complicated models, such as random forests.
```

Comment:
Why mention the logarithm of the population here? You should either explain what that would accomplish, or remove this comment from the conclusion.

Also, I expected the skewness of the income and the house values to have come in play in the algorithm as well.

I though you'd have used log(1 + p) on both of those features and the applied the inverse function for the final prediction. I think either we do that, or just don't comment on the skewness at all.

Reply:
Ok, I will remove the logarithm of the population from the conclusion because it doesn't make sense.  I also tried with log(1+p) of mentioned features and it turns out to make prediction worse than without logs. However, I would like to keep the mention of skewness. I made it mostly because of the correlation matrix at the end of the notebook.